### PR TITLE
Split vision.py into smaller scripts

### PR DESCRIPTION
**Pull request by rahul-tuli**
Refactor: vision.py into 4 scripts
* train.py for training specific tasks
* lr_analysis.py for lr sensitivity analysis
* pr_analysis.py for Kernel Sparsity (pruning) analysis
* export.py for export a model to onnx and store input/outputs

Added:
- Quality fixes for function arguments and * imports
- Scripts tested locally with example commands from old vision.py
- vision.py modified to point user to new scripts
- updation of README
- updation of example commands in recipes
- NmArgumentParser class for parsing dataclasses into arguments
- Tests to show example usage of the parser
- Support for both hyphenated and non-hyphenated arguments

The scripts are tested locally using the following commands on `imagenette` dataset
```python
python integrations/pytorch/train.py \
    --recipe-path integrations/pytorch/recipes/resnet50-pruned.md \
    --arch-key resnet50 --dataset imagenette --dataset-path ~/datasets/ \
    --train-batch-size 256 --test-batch-size 1024


python integrations/pytorch/export.py \
    --arch-key resnet50 --dataset imagenette --dataset-path ~/datasets/



python integrations/pytorch/pr_sensitivity.py \
    --approximate --arch-key resnet50 --dataset imagenette \
    --dataset-path ~/datasets/


python integrations/pytorch/lr_analysis.py \
    --arch-key mobilenet --dataset imagenette \
    --dataset-path ~/datasets/ --batch-size 2
```

### CODE REVIEW

File: integrations/pytorch/train.py
Comment (lucaslie) [inline integrations/pytorch/train.py:155]: I am not 100% familiar with the import structure in `sparseml` but would this import work if one tried to execute the script from another directory? 
Or is that not supported?

File: integrations/pytorch/export.py
Comment (lucaslie) [inline integrations/pytorch/export.py:18]: are those up-to-date? Not sure if the argparser changed much because of the split but might be worth double-checking?

**Review by lucaslie — COMMENTED** [review]
I checked over it and left suggestions for a few small changes. But I didn't run it to verify that it's working.

**Comment by markurtz** [comment]
Putting a note in here, @rahul-tuli please wait to land this until I have some time to look through given the size of the change and impact on testing. I'll try to get through it over the next few days, thanks

**Comment by rahul-tuli** [comment]
Thanks @lucaslie taking a deeper look everywhere for references to vision.py

File: integrations/pytorch/train.py
Comment (rahul-tuli) [reply integrations/pytorch/train.py]: Resolved! Thanks for catching these, don't know how missed them somehow.

File: integrations/pytorch/recipes/mobilenet-pruned_quant.md
Comment (rahul-tuli) [reply integrations/pytorch/recipes/mobilenet-pruned_quant.md:148]: Resolved!

File: integrations/pytorch/recipes/resnet50-imagenette-pruned.md
Comment (rahul-tuli) [reply integrations/pytorch/recipes/resnet50-imagenette-pruned.md:133]: Resolved!

File: integrations/pytorch/recipes/resnet50-pruned.md
Comment (rahul-tuli) [reply integrations/pytorch/recipes/resnet50-pruned.md:73]: Resolved!

File: integrations/pytorch/recipes/resnet50-pruned_quant.md
Comment (rahul-tuli) [reply integrations/pytorch/recipes/resnet50-pruned_quant.md:186]: resolved!

File: integrations/pytorch/train.py
Comment (rahul-tuli) [reply integrations/pytorch/train.py:155]: It should work when executed from a different directory @lucaslie 
```python
import sys
print(sys.path)
```

The train, export, analysis scripts have '/home/rahul/projects/sparseml/integrations/pytorch' in their `sys.path` so python should be able to find `helpers.py`, it would work as long as helpers is not in a separate directory. 

Try running the following command

```bash
python sparseml/integrations/pytorch/analysis.py lr_sensitivity \
    --arch-key mobilenet --dataset imagenette \
    --dataset-path ~/datasets --batch-size 2
```

It takes about ~20secs to run. It works on my machine, let me know if you get an error, will change this to a relative import then if needed.

File: integrations/pytorch/export.py
Comment (rahul-tuli) [reply integrations/pytorch/export.py:18]: Yes, those are up to date, generated fresh using --help

**Comment by rahul-tuli** [comment]
> Putting a note in here, @rahul-tuli please wait to land this until I have some time to look through given the size of the change and impact on testing. I'll try to get through it over the next few days, thanks

Sure @markurtz one thing I wanted your take on is:
My Two train of thoughts
1.  Should the arguments be in the same script where they are needed. Will add code duplicacy but a dev-user won't have to go to a separate file to check/change/add arguments.

2. The current implementation removes code duplicacy but a dev-user might have to go to helpers.py to check/change/add arguments

If 1 is preferred over 2, let me know I have that implementation in another local branch.

File: integrations/pytorch/README.md
Comment (rahul-tuli) [reply integrations/pytorch/README.md]: Done!

File: integrations/pytorch/recipes/mobilenet-pruned.md
Comment (rahul-tuli) [reply integrations/pytorch/recipes/mobilenet-pruned.md:100]: fixed!

Comment (rahul-tuli) [reply integrations/pytorch/recipes/mobilenet-pruned.md:100]: done!

File: integrations/pytorch/README.md
Comment (lucaslie) [inline integrations/pytorch/README.md:90]: sorry to be nit-picky about it but I would suggest hyper-referencing relative paths, e.g.:

```
 [notebooks](./integrations/pytorch/notebooks)
```

This way it will auto-resolve to the correct branch depending on the branch you are on and IDEs like VSCode can also link to the right directory instead of going to the main branch on github.com

File: integrations/pytorch/train.py
Comment (lucaslie) [reply integrations/pytorch/train.py:155]: Sounds good. I will check it before submitting my review.

**Comment by lucaslie** [comment]
> > Putting a note in here, @rahul-tuli please wait to land this until I have some time to look through given the size of the change and impact on testing. I'll try to get through it over the next few days, thanks
> 
> Sure @markurtz one thing I wanted your take on is:
> My Two train of thoughts
> 
>     1. Should the arguments be in the same script where they are needed. Will add code duplicacy but a dev-user won't have to go to a separate file to check/change/add arguments.
> 
>     2. The current implementation removes code duplicacy but a dev-user might have to go to helpers.py to check/change/add arguments
> 
> 
> If 1 is preferred over 2, let me know I have that implementation in another local branch.

So I was brain-storming a bit and checking over the current set of functionality in `helpers.py`. I was also looking at pytorch's [vision/references](https://github.com/pytorch/vision/tree/master/references), e.g., the [classification reference](https://github.com/pytorch/vision/tree/master/references/classification).

They definitely seem to copy/paste a lot of code across the different reference implementations. But they also have a very lean and short `utils.py` in most sub-folders solely dedicated to the folder. 

I like this model and I think that we should follow it by using `helpers.py` (or renamed as `utils.py`?) as a way to abstract away irrelevant details but also keep it simple. 

So some things we could do to simplify the dependency on `helpers.py` without copy/pasting all the code: 
* Simplify how to argparsers are created. Maybe just one common set of args plus one script-specific `add_args`
* remove global variables in `helpers.py` that we then import 
* remove the `CURRENT_TASK` in the scripts and the corresponding if, else statements in the helper functions. If it doesn't generalize without if, else let's just keep it separate
* remove the global variable for `LOGGER` in `helpers.py` 

The list is probably not complete but hopefully captures the spirit of what I am trying to say.

On a more general note: I am not a fan of enumerating function imports from a module like `helpers` instead of explicitly importing the entire module and then calling it explicitly like `helpers.add_universal_args`. But if that's the style that we generally adopt, please feel free to ignore my comment.

File: integrations/pytorch/train.py
Comment (lucaslie) [reply integrations/pytorch/train.py:155]: yes, that works :)

File: integrations/pytorch/README.md
Comment (rahul-tuli) [reply integrations/pytorch/README.md:90]: Thanks @lucaslie didn't know about this


nit
`[notebooks](notebooks)`

**Comment by rahul-tuli** [comment]
Thanks @lucaslie those are very valid points that you've put forth. Will update my design accordingly. Really appreciate the effort. 😇

**Review by bfineran — COMMENTED** [review]
Great work @rahul-tuli! Changes look overall good to me, thanks for the thorough update to the script references.

Two things:

1. make sure to test `train.py` using DDP with multiple GPUs (let's sync offline about this)
2. Since the `add_*_args()` functions have a notion of `task` anyways, what is the reasoning for needing to split them out into multiple functions to be called by each task? If we're going to keep the argparse definition in a separate place, it may be nicer to keep it created in a single function, that way a reader will be able to quickly see which args go to which tasks, rather than having to switch between files.

**Comment by rahul-tuli** [comment]
> Great work @rahul-tuli! Changes look overall good to me, thanks for the thorough update to the script references.
> 
> Two things:
> 
> 1. make sure to test `train.py` using DDP with multiple GPUs (let's sync offline about this)
> 2. Since the `add_*_args()` functions have a notion of `task` anyways, what is the reasoning for needing to split them out into multiple functions to be called by each task? If we're going to keep the argparse definition in a separate place, it may be nicer to keep it created in a single function, that way a reader will be able to quickly see which args go to which tasks, rather than having to switch between files.


I like this train of thought. Let's sync again offline about the final design. Wil push out changes accordingly. :)

**Review by lucaslie — CHANGES_REQUESTED** [review]
Hi @rahul-tuli,

I did another pass over the PR and I really like the structure now. :) Very intuitive for somebody who sees it literally for the first time lol. 

One general question I had, which is probably not related to this particular PR but might be worth considering: 

it seems that the log/results directories that are automatically created during a run (e.g. `pytorch_vision_train`) are not captured by our `.gitignore` and so it pops up as untracked files. Maybe it would be nice to include them in our `.gitignore`?

**Review by bfineran — APPROVED** [review]
LGTM, make sure to test a simple DDP flow in train.py before landing

**Review by markurtz — CHANGES_REQUESTED** [review]
This is looking good Rahul. I have a few comments throughout and two general ones:
- Let's remove the passing of namespace into the functions. It makes it easier to code now, but makes it harder to debug and track variable state later. An alternative would be using typed classes for that to make it clear what can and cannot be passed through to each function.
- Please be sure to resolve all comments that have been fixed and are no longer open or in discussion

**Review by lucaslie — CHANGES_REQUESTED** [review]
just another few, minor comments.

**Review by bfineran — DISMISSED** [review]
should have caught this before, only train.py needs DDP support

File: integrations/pytorch/analysis.py
Comment (rahul-tuli) [reply integrations/pytorch/analysis.py:321]: Done!

File: integrations/pytorch/utils.py
Comment (rahul-tuli) [reply integrations/pytorch/utils.py:448]: Done!

File: integrations/pytorch/export.py
Comment (rahul-tuli) [reply integrations/pytorch/export.py:215]: Thanks! changed

File: integrations/pytorch/analysis.py
Comment (rahul-tuli) [reply integrations/pytorch/analysis.py]: Changed!

File: integrations/pytorch/analysis.py
Comment (rahul-tuli) [reply integrations/pytorch/analysis.py]: Pulled out to a separate script

File: integrations/pytorch/train.py
Comment (rahul-tuli) [reply integrations/pytorch/train.py]: Added

File: integrations/pytorch/analysis.py
Comment (rahul-tuli) [reply integrations/pytorch/analysis.py]: No more for loop

**Review by markurtz — APPROVED** [review]

**Review by lucaslie — APPROVED** [review]
looks great. Great job!

Final comment: do we need to add Huggingface's licence to the code since our `NmArgumentParser` very much resembles their argparser class or is the current reference `# Inspired from ...` enough?

**Comment by rahul-tuli** [comment]
> looks great. Great job!
> 
> Final comment: do we need to add Huggingface's licence to the code since our `NmArgumentParser` very much resembles their argparser class or is the current reference `# Inspired from ...` enough?

We should be good cause we already refer to their license in our notice [here](https://github.com/neuralmagic/sparseml/blob/fe5d49cc2f116b29f25d5ef10c2ae048c13d3b82/NOTICE#L37)

**Comment by rahul-tuli** [comment]
> should have caught this before, only train.py needs DDP support

Dismissing review cause it's stale!

**Pull request merged**
